cmd/callVariants/testVariant.go

```diff
@@ -0,0 +1,359 @@
+		sort.Slice(toSort, func(i, j int) bool {
+			if len(toSort[i]) == len(toSort[j]) {
+				return toSort[i] < toSort[j]
+			}
+			return len(toSort[i]) < len(toSort[j])
+		})
+	}
+
+	return possibleAlts, isInsertion
+}
+
+// getPassingAlts returns a slice of strings with all alternate alleles with at least 1 samples with a significant p values
+// as well as a slice of p values for all samples for each passing alternate allele.
+func getPassingAlts(possibleAlts []string, altPvalues [][]float64, isInsertion []bool, maxP float64) (passingAlts []string, passingAltPvalues [][]float64, passingIsInsertion []bool) {
+	var pass bool
+	passingIsInsertion = make([]bool, 0, len(isInsertion))
+	for i := range altPvalues { // first slice is possibleAlts, second slice is experimental samples
+		pass = false
+		for j := range altPvalues[i] {
+			if altPvalues[i][j] < maxP {
+				pass = true
+				break
+			}
+		}
+		if pass {
+			passingAlts = append(passingAlts, possibleAlts[i])
+			passingAltPvalues = append(passingAltPvalues, altPvalues[i])
+			passingIsInsertion = append(passingIsInsertion, isInsertion[i])
+		}
+	}
+	return
+}
+
+// fishersExactTest returns the p value for the alternate allele present in the exp pile, based on the bkgd pile.
+func fishersExactTest(altString string, exp sam.Pile, bkgd sam.Pile, hasNorm bool, minAf float64, minCoverage int, isInsertion bool) float64 {
+	// Begin gathering parameters for Fishers Exact Test done in the numbers package
+	// test is for the matrix:
+	// [a b]
+	// [c d]
+	// a = Experimental Depth
+	// b = Background Depth
+	// c = Experimental Alt Allele Count
+	// d = Background Alt Allele Count
+
+	var a, b, c, d int
+
+	if !isInsertion { // alt in single base
+		alt := int(dna.RuneToBase(rune(altString[0])))
+		c = exp.Count[alt]
+		d = bkgd.Count[alt]
+	} else { // alt in insertion
+		c = exp.InsCount[altString]
+		d = bkgd.InsCount[altString]
+	}
+
+	a = calcDepth(exp) - c
+	b = calcDepth(bkgd) - d
+
+	// if no normal is present then bkgd is the sum of all experimental bases,
+	// therefore the current experimental base must be subtracted from bkgd totals.
+	if !hasNorm {
+		b -= a
+		d -= c
+	}
+
+	// Check exclusion cases to avoid actually doing the test
+	var p float64
+	switch {
+	// If alternate allele is zero then there is no variant and score is 1
+	case c == 0:
+		p = 1
+
+	// If a = b and c = d then it is testing itself and should return 1
+	case a == b && c == d:
+		p = 1
+
+	// If the allele frequency of d > c then p is 1
+	case float64(c)/float64(c+a) < float64(d)/float64(d+b):
+		p = 1
+
+	// If the coverage is less then minCoverage then p is noted as 1 so as to be excluded
+	case a+c < minCoverage:
+		p = 1
+
+	// If the allele frequency is less than the threshold then p is noted as 1 so as to be excluded
+	case float64(c)/float64(c+a) < minAf:
+		p = 1
+
+	// If no exclusion conditions are met, then calculate p value
+	default:
+		if a < c && b < d { // alt is common allele
```

Comment:
Do we need to multiply these pvalues by 2 since we are checking to see if a lower tail or upper tail test would give a more extreme value (essentially doing both tests and reporting the one that is most extreme).

Reply:
This was added when I was debugging, but I found a different bug which made it unnecessary. It is now changed so aSmall is always true.